common/nexus/incoming_service_registry.go

```diff
@@ -0,0 +1,355 @@
+			// Don't let this get near our call timeout, otherwise we can't tell the difference
+			// between a fast reply and a timeout.
+			minWaitTime = min(minWaitTime*2, r.config.refreshLongPollTimeout()/2)
+		} else {
+			minWaitTime = r.config.refreshMinWait()
+		}
+	}
+
+	return ctx.Err()
+}
+
+// refreshServices sends long-poll requests to matching to check for any updates to service data.
+// It waits to send any requests until service data has been initialized.
+func (r *IncomingServiceRegistry) refreshServices(ctx context.Context) error {
+	// Wait for service data to be initialized before long polling.
+	if _, err := r.serviceDataReady.Get(ctx); err != nil {
+		return err
+	}
+
+	r.dataLock.RLock()
+	currentTableVersion := r.tableVersion
+	r.dataLock.RUnlock()
+
+	resp, err := r.matchingClient.ListNexusIncomingServices(ctx, &matchingservice.ListNexusIncomingServicesRequest{
+		NextPageToken:         nil,
+		PageSize:              int32(r.config.refreshPageSize()),
+		LastKnownTableVersion: currentTableVersion,
+		Wait:                  true,
+	})
+	if err != nil {
+		return err
+	}
+
+	if resp.TableVersion == currentTableVersion {
+		// Long poll returned with no changes
+		return nil
+	}
+
+	// changes were returned by long poll so acquire write lock to prevent reads while updating
+	r.dataLock.Lock()
+	defer r.dataLock.Unlock()
+
+	if r.tableVersion >= resp.TableVersion {
+		// service data was updated while waiting for the write lock
+		return nil
+	}
+
+	currentTableVersion = resp.TableVersion
+	services := make(map[string]*nexus.IncomingService)
+	for _, service := range resp.Services {
+		services[service.Id] = service
+	}
+
+	currentPageToken := resp.NextPageToken
+	for currentPageToken != nil {
+		resp, err = r.matchingClient.ListNexusIncomingServices(ctx, &matchingservice.ListNexusIncomingServicesRequest{
+			NextPageToken:         currentPageToken,
+			PageSize:              int32(r.config.refreshPageSize()),
+			LastKnownTableVersion: currentTableVersion,
+			Wait:                  false,
+		})
+
+		if err != nil {
+			if errors.Is(err, p.ErrNexusTableVersionConflict) {
+				// Indicates table was updated during paging, so reset and start from the beginning
+				currentTableVersion, services, err = r.getAllServicesMatching(ctx)
+				if err != nil {
+					return err
+				}
+				break
+			}
+			return err
+		}
+
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		for _, service := range resp.Services {
+			services[service.Id] = service
+		}
+
+		currentPageToken = resp.NextPageToken
+	}
+
+	r.tableVersion = currentTableVersion
+	r.services = services
+	return nil
+}
+
+// initializeServices initializes the in-memory view of service data.
+// It first tries to load from matching service and falls back to querying persistence directly if matching is unavailable.
+func (r *IncomingServiceRegistry) initializeServices(ctx context.Context) error {
+	r.dataLock.Lock()
+	defer r.dataLock.Unlock()
```

Comment:
This is slightly troubling, if two requests try to initialize the services and one has a shorter context timeout, it'll possibly stay blocked for the duration of the longer request's timeout.

Reply:
That's true. But I'm not sure there's an easy way to avoid that. We can't have two threads trying to initialize the data at the same time.

Reply:
You can wait on the lock in a goroutine that closes a channel on completion and select on that channel and the context done channel. It adds some additional overhead for every request, but that can be eliminated by checking `r.serviceDataReady.Ready` before trying to acquire the lock.